File: Assign1/Go/Src/Concurrency/MyProblem.go
package main
import (
	"fmt"
	"sync"
	"time"
)

var input string
func userReadeThread(m *sync.RWMutex, id int) {
	for {
		m.RLock()
		fmt.Println("User", id, " is reading");
		m.RUnlock()
	}

}

func userWriteThread(m *sync.RWMutex, id int, wg *sync.WaitGroup, times *int) {

	for{
		m.Lock()
		fmt.Println("User", id, " starts to write");
		fmt.Printf("Please input your information: ")
		fmt.Scanln(&input)
		fmt.Println("User", id, " starts to write to database")
		*times++;
		if *times == 10{
			wg.Done()
		}
		m.Unlock()
	}

}

func main() {
	t1 := time.Now()
	var wg sync.WaitGroup
	var m sync.RWMutex
	times := 0

	wg.Add(1)

	for i := 0; i < 10; i++ {
		go userReadeThread(&m, i)
	}

	for i := 0; i < 3; i++{
		go userWriteThread(&m, i, &wg, &times)
	}

	wg.Wait()

	elapsed := time.Since(t1)
	fmt.Println("All threads elapsed: ", elapsed)

}

File: Assign1/Java/Concurrency/src/MyProblem.java

import java.util.Scanner;
import java.util.concurrent.locks.ReentrantReadWriteLock;

import javax.swing.JTextField;



public class MyProblem {
	private ReentrantReadWriteLock lock = new ReentrantReadWriteLock();   
	
	private static int loop;
	private static long startTime;

    public static void main(String[] args) {    	
    	MyProblem mp = new MyProblem();
    	loop = 0;
    	startTime=System.currentTimeMillis();  

    	for(int i = 0; i < 3; i++)
        {
        	new Thread(mp.new Writer(i)).start();
        }
        for(int i = 0; i < 10; i++)
        {
        	
	       	new Thread(mp.new Reader(i)).start();
        }
    }
  
    
    class Writer implements Runnable {
    	private int  id;    	
    	JTextField text;
        public Writer(int id)
        {
            this.id = id;
        }
        public void run()  
        {     
        	while(true)
        	{   		
	        	try {
	                Thread.sleep(1);
	            } catch (InterruptedException e1) {
	                e1.printStackTrace();
	            } 	        	
	        	lock.writeLock().lock();      
	        	System.out.println("please input your message:");
	    		Scanner sc = new Scanner(System.in);	
	            System.out.println("User" + id + "input is :" + sc.nextLine()); 
	            loop++;
	            if (loop == 10)
	            {
	            	long endTime=System.currentTimeMillis();       
	                System.out.println("All threads elapsed: "+(endTime-startTime)+"ms");
	            }
	            lock.writeLock().unlock();  
        	}
        }  
    }
  
    class Reader implements Runnable {   
    	private int  id;
    	
        public Reader(int id)
        {
            this.id = id;
        }

    	public void run()  
        {   
    		while(true)
    		{
		    	try {
		           Thread.sleep(1);
		        } catch (InterruptedException e1) {
		           e1.printStackTrace();
		        }		
		    	lock.readLock().lock();
		        System.out.println("User " + id + " is reading"); 
		        lock.readLock().unlock();
    		}
        } 
    }
}

